### PR TITLE
Remove integration test dependency on commercial app-interface

### PR DESCRIPTION
**Pull request by dwelch0**
In order to test functionality being developed for [vault fedramp epic](https://issues.redhat.com/browse/APPSRE-4655), the existing test environment requires significant alteration. The primary shortcoming of existing test environment is the reliance on commercial app-interface/qontract-server when performing queries within bats tests. Additionally, in order to test new oidc functionality, a keycloak instance is required.

This PR adds a minimal app-interface config, containing data only relevant to vault-manager testing. With new approach, qontract-server and keycloak are deployed and targeted by vault-manager/bats when performing tests.

Notes: 
- elected to generate (bundle) data.json and commit to repo instead of incorporating validator image and running on each build to reduce build time

### CODE REVIEW

File: tests/app-interface/data/services/vault/config/auth-backends/oidc.yml [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+_path: "oidc/"
+type: "oidc"
+description: "rh sso auth backend"
+settings:
+  config:
+    _type: "oidc"
+    oidc_discovery_url: "http://localhost:8180/auth/realms/test"
+    oidc_client_id: "vault"
+    oidc_client_secret: "my-special-client-secret"
+    default_role: "default"
```
Comment (TGPSKI) [inline tests/app-interface/data/services/vault/config/auth-backends/oidc.yml:16]: newline

File: tests/app-interface/data/services/vault/config/roles/oidc/minimal.yml [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  token_ttl: "60"
+  token_max_ttl: "60"
+  token_explicit_max_ttl: "0"
+  token_type: "default"
+  token_period: "0"
+  token_policies: []
+  token_bound_cidrs: []
+  token_no_default_policy: false
+  token_num_uses: "0"
+  user_claim: "email"
```
Comment (TGPSKI) [inline tests/app-interface/data/services/vault/config/roles/oidc/minimal.yml:24]: newline

File: tests/app-interface/docs/context.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
+# App-Interface
```
Comment (TGPSKI) [inline tests/app-interface/docs/context.md:1]: Clean this up and include info on why this is here and how you build data.json. Nothing else

File: tests/app-interface/hack/toc.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
+#!/usr/bin/env python
```
Comment (TGPSKI) [inline tests/app-interface/hack/toc.py:1]: you can get rid of this

File: tests/bats/audit/audit-devices.bats [hunk 5]
```diff
@@ -8,6 +8,7 @@ load ../helpers
     #
     export GRAPHQL_QUERY_FILE=/tests/fixtures/audit/enable_audit_device.graphql
     run vault-manager
+    echo $output
```
Comment (TGPSKI) [inline tests/bats/audit/audit-devices.bats:11]: is this change for all test files or just these two? keep it consistent if you are adding this as a test debug feature

File: tests/bats/helpers.bash [hunk 6]
```diff
@@ -2,7 +2,12 @@ setup() {
   export VAULT_ADDR=http://127.0.0.1:8200
   export VAULT_TOKEN=root
   export VAULT_AUTHTYPE=token
-  docker run -d --net=host --name="vault-dev-server" --cap-add=IPC_LOCK -e 'VAULT_DEV_ROOT_TOKEN_ID=root' -p 8200:8200 -v /tmp/:/var/log/vault/:Z quay.io/app-sre/vault:1.5.4
+  docker run -d --name="vault-dev-server" \
```
Comment (TGPSKI) [inline tests/bats/helpers.bash:5]: when you add the cleanup flow controls to the bash script, can you eliminate this setup() method entirely?

File: tests/run-tests.sh [hunk 7]
```diff
@@ -2,7 +2,36 @@
 
 set -e
 
+# docker rm -f qontract-server
```
Comment (TGPSKI) [inline tests/run-tests.sh:5]: remove

File: tests/keycloak/realm-client.json [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+        "secret": "my-special-client-secret",
+        "redirectUris": [
+          "*"
+        ],
+        "webOrigins": [
+          "*"
+        ]
+      }
+    ]
+  }
```
Comment (TGPSKI) [inline tests/keycloak/realm-client.json:20]: newline

File: tests/keycloak/user.json [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+        "firstName": "Test",
+        "lastName": "McTester",
+        "attributes": {
+          "locale": [
+            "de"
+          ]
+        }
+      }
+    ]
+  }
```
Comment (TGPSKI) [inline tests/keycloak/user.json:18]: newline

File: tests/run-tests.sh [hunk 10]
```diff
@@ -2,7 +2,40 @@
...
+cd app-interface && make server || error_cleanup
+cd ..
+
 for test in $(find bats/ -type f | grep .bats); do
     echo "running $test"
     bats --tap $test
 done
+
+docker rm -f qontract-server || true
+docker rm -f keycloak || true
```
Comment (TGPSKI) [inline tests/run-tests.sh:41]: newline

File: tests/run-tests.sh [hunk 11]
```diff
@@ -2,7 +2,40 @@
 
 set -e
 
+error_cleanup () {
+  docker rm -f keycloak || true
+  docker rm -f qontract-server || true
+  echo "Docker environment cleaned"
+  exit 1
+}
+
```
Comment (TGPSKI) [inline tests/run-tests.sh:11]: cleanup() {
  docker rm -rf vault || true
  docker rm -rf keycloak || true
  docker rm -rf qontract-server |\ true
  echo "Docker environment cleaned"
  echo 0
}

File: tests/run-tests.sh [hunk 12]
```diff
@@ -2,7 +2,40 @@
...
+cd app-interface && make server || error_cleanup
+cd ..
+
 for test in $(find bats/ -type f | grep .bats); do
     echo "running $test"
     bats --tap $test
 done
+
+docker rm -f qontract-server || true
```
Comment (TGPSKI) [inline tests/run-tests.sh:40]: just call cleanup here

File: tests/run-tests.sh [hunk 13]
```diff
@@ -2,7 +2,40 @@
...
+
+docker run -d --name="keycloak" \
+    --net=host --cap-add=IPC_LOCK \
+    -e KEYCLOAK_ADMIN=admin -e KEYCLOAK_ADMIN_PASSWORD=admin \
+    -p 8180:8180 \
+    quay.io/keycloak/keycloak:17.0 \
+    start-dev \
+    --http-port 8180 \
+    --http-relative-path /auth || error_cleanup
+
```
Comment (TGPSKI) [inline tests/run-tests.sh:20]: you have to wait to make sure the server responds before continuing

File: tests/app-interface/Makefile [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+		$(VALIDATOR_IMAGE):$(VALIDATOR_IMAGE_TAG) \
+		qontract-bundler /schemas /graphql/schema.yml /data /resources placeholder placeholder > $(OUTPUT_DIR)/$(BUNDLE_FILENAME)
+
+validate:
+	@$(CONTAINER_ENGINE) run --rm \
+		-v $(OUTPUT_DIR):/bundle:z \
+		$(VALIDATOR_IMAGE):$(VALIDATOR_IMAGE_TAG) \
+		qontract-validator --only-errors /bundle/$(BUNDLE_FILENAME)
+
+toc:
```
Comment (TGPSKI) [inline tests/app-interface/Makefile:33]: remove

File: tests/app-interface/Makefile [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+	./hack/toc.py
+
+run:
```
Comment (TGPSKI) [inline tests/app-interface/Makefile:36]: you could remove this and run this exactly the same way you are setting up vault, keycloak, and keycloak-config-cli

seems like the right thing. keep the calling of docker containers consistent

File: tests/app-interface/Makefile [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+	@$(CONTAINER_ENGINE) run -d -it --rm \
+		--name=qontract-server \
+		-v $(HOST_PATH)/$(OUTPUT_DIR):/bundle:z \
+		-p 4000:4000 \
+		-e LOAD_METHOD=fs \
+		-e DATAFILES_FILE=/bundle/$(BUNDLE_FILENAME) \
+		$(QONTRACT_SERVER_IMAGE):$(QONTRACT_SERVER_IMAGE_TAG)
+
+dev: bundle validate run
```
Comment (TGPSKI) [inline tests/app-interface/Makefile:45]: remove

File: tests/app-interface/Makefile [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+
+server: run
```
Comment (TGPSKI) [inline tests/app-interface/Makefile:47]: remove

File: tests/Dockerfile.tests [hunk 18]
```diff
@@ -3,7 +3,7 @@ COPY . .
 RUN make gobuild
 
 FROM quay.io/centos/centos:7
-ENV PACKAGES="jq git docker unzip wget"
+ENV PACKAGES="jq git docker unzip wget make"
```
Comment (TGPSKI) [inline tests/Dockerfile.tests:6]: you can remove make from env packages if you move the docker command to the same place as the other docker container starts.

you can flow control errors the same way too

File: tests/bats/helpers.bash (hunk 6 above)
Comment (TGPSKI) [reply tests/bats/helpers.bash:5]: also you can remove teardown below as well

File: tests/run-tests.sh [hunk 19]
```diff
@@ -2,7 +2,105 @@
...
+    -v $HOST_PATH/$(pwd)/keycloak:/config \
+    $KEYCLOAK_CLI_IMAGE:$KEYCLOAK_CLI_TAG
+
+docker run -d -it --rm \
+		--name=$QONTRACT_SERVER_NAME \
+		-v $HOST_PATH/$(pwd)/app-interface:/bundle:z \
+		-p 4000:4000 \
+		-e LOAD_METHOD=fs \
+		-e DATAFILES_FILE=/bundle/data.json \
+		$QONTRACT_SERVER_IMAGE:$QONTRACT_SERVER_IMAGE_TAG
```
Comment (TGPSKI) [inline tests/run-tests.sh:87]: container alive for qontract-server

File: tests/run-tests.sh [hunk 20]
```diff
@@ -2,7 +2,105 @@
...
+    start-dev \
+    --http-port 8180 \
+    --http-relative-path /auth \
+    --metrics-enabled=true
+container_alive "http://127.0.0.1:8180/auth"
+
+docker run --name $KEYCLOAK_CLI_NAME \
+    --net=host \
+    -e KEYCLOAK_URL="http://localhost:8180/auth" \
+    -e KEYCLOAK_USER=$KEYCLOAK_USER \
```
Comment (TGPSKI) [inline tests/run-tests.sh:73]: use consistent whitespace for all docker commands

File: tests/run-tests.sh [hunk 21]
```diff
@@ -2,7 +2,107 @@
...
+      cleanup
+      exit 1
+    fi
+    ((++idx))
+  done
+}
+
+cleanup
+
+docker run -d --name=$KEYCLOAK_NAME \
```
Comment (TGPSKI) [inline tests/run-tests.sh:58]: add comments for clarity on each block to let us know whats happening

# Spin up keycloak server

or something

File: tests/run-tests.sh [hunk 22]
```diff
@@ -2,7 +2,107 @@
...
+  until $(curl --output /dev/null --silent --head --fail $1); do
+    printf '.'
+    sleep 1
+    if [[ $idx == $2 ]]; then
+      cleanup
+      exit 1
+    fi
+    ((++idx))
+  done
+}
```
Comment (TGPSKI) [inline tests/run-tests.sh:54]: add lines between done and }

echo ""
echo "connectivty established with $3"
echo ""

container_alive xxx $CONTAINER_HEALTH_TIMEOUT_DEFAULT $QONTRACT_SERVER_NAME

or whatever name is being run

Comment (TGPSKI) [reply tests/run-tests.sh:54]: also for completeness do the beginning too

File: tests/run-tests.sh [hunk 23]
```diff
@@ -2,7 +2,107 @@
...
+    docker rm -f $QONTRACT_SERVER_NAME
+  fi
+  if docker ps -a --format "table {{.Names}}" | grep -qw $VAULT_NAME; then
+    docker rm -f $VAULT_NAME
+  fi
+  echo "Docker environment cleaned"
+}
+
+container_alive () {
+  idx=0
```
Comment (TGPSKI) [inline tests/run-tests.sh:44]: echo ""
echo "testing connectivity with container $3"
echo ""

File: tests/run-tests.sh [hunk 24]
```diff
@@ -2,7 +2,107 @@
...
+  until $(curl --output /dev/null --silent --head --fail $1); do
+    printf '.'
+    sleep 1
+    if [[ $idx == $2 ]]; then
+      cleanup
```
Comment (TGPSKI) [inline tests/run-tests.sh:49]: echo ""
echo ""connectivity to container $3 failed"
echo ""

File: tests/run-tests.sh [hunk 25]
```diff
@@ -2,7 +2,122 @@
...
+  -e KEYCLOAK_USER=$KEYCLOAK_USER \
+  -e KEYCLOAK_PASSWORD=$KEYCLOAK_PASSWORD \
+  -e KEYCLOAK_AVAILABILITYCHECK_ENABLED=true \
+  -e KEYCLOAK_AVAILABILITYCHECK_TIMEOUT=120s \
+  -e IMPORT_FILES='/config/*' \
+  -v $HOST_PATH/$(pwd)/keycloak:/config \
+  $KEYCLOAK_CLI_IMAGE:$KEYCLOAK_CLI_TAG
+
+# spin up qontract-server, using existing data.json file
+docker run -d -it --rm \
```
Comment (TGPSKI) [inline tests/run-tests.sh:94]: we dont need -it on this call I think

File: tests/run-tests.sh [hunk 26]
```diff
@@ -2,7 +2,122 @@
...
+docker run -d --name=$KEYCLOAK_NAME \
+  --net=host \
+  --cap-add=IPC_LOCK \
+  -e KEYCLOAK_ADMIN=$KEYCLOAK_USER -e KEYCLOAK_ADMIN_PASSWORD=$KEYCLOAK_PASSWORD \
+  -p 8180:8180 \
+  $KEYCLOAK_IMAGE:$KEYCLOAK_IMAGE_TAG \
+  start-dev \
+  --http-port 8180 \
+  --http-relative-path /auth \
+  --metrics-enabled=true
```
Comment (TGPSKI) [inline tests/run-tests.sh:78]: do we need metrics if we aren't using the /health endpoint

Comment (dwelch0) [reply tests/run-tests.sh:78]: nope. good catch. removing

File: tests/run-tests.sh (hunk 25 above)
Comment (dwelch0) [reply tests/run-tests.sh:94]: it existed in makefile definition but questioned it as well. will remove

**Pull request merged**